### PR TITLE
Resizable UIWindow

### PR DESCRIPTION
**Pull request by ephemer**
**Type of change:** Graphical improvement

## Motivation (current vs expected behavior)

We want to make UIWindow resizable depending on external factors. For example - by resizing the window on Mac, or by showing/hiding the status bar on Android, entering/leaving immersive mode etc.

## Please check if the PR fulfills these requirements
- [ ] Self-review: I am confident this is the simplest and clearest way to achieve the expected behaviour
- [ ] There are no dependencies on other PRs or I have linked dependencies through Zenhub
- [ ] The commit messages are clean and understandable
- [ ] Tests for the changes have been added (for bug fixes / features)

### CODE REVIEW

File: Sources/UIScreen.swift [hunk 1]
```diff
@@ -30,9 +30,14 @@ public final class UIScreen {
         self.rawPointer = renderTarget
         self.bounds = bounds
         self.scale = scale
+
+        if let window = UIApplication.shared?.keyWindow {
+            window.frame = bounds
```
Comment (rikner) [inline Sources/UIScreen.swift:35]: window.rootViewController.viewWillTransition(...)`

File: Sources/UIScreen.swift [hunk 2]
```diff
@@ -25,7 +25,20 @@ public final class UIScreen {
     // Keep that in mind when using it: i.e. if possible, don't ;)
     internal var rawPointer: UnsafeMutablePointer<GPU_Target>!
 
-    nonisolated public let bounds: CGRect
+    public var bounds: CGRect {
+        didSet {
+            GPU_SetWindowResolution(
+                UInt16(bounds.width),
+                UInt16(bounds.height)
```
Comment (rikner) [inline Sources/UIScreen.swift:32]: `.rounded()`

File: Sources/CALayer.swift [hunk 3]
```diff
@@ -92,7 +92,7 @@ open class CALayer {
             )
 
             guard let inverseTransform = affineTransform().inverted() else {
-                assertionFailure("You tried to set the frame of a CALayer whose transform cannot be inverted. This is undefined behaviour.")
+//                assertionFailure("You tried to set the frame of a CALayer whose transform cannot be inverted. This is undefined behaviour.")
```
Comment (michaelknoch) [inline Sources/CALayer.swift:95]: revert this change? @rikner

File: Sources/UIAlertController.swift [hunk 4]
```diff
@@ -63,6 +63,12 @@ public class UIAlertController: UIViewController {
         )
     }
 
+    public override func viewWillLayoutSubviews() {
+        let size = view.bounds.size
+        self.alertControllerView?.center = CGPoint(x: size.width / 2,y: size.height / 2)
+        super.viewWillLayoutSubviews()
```
Comment (michaelknoch) [inline Sources/UIAlertController.swift:69]: is this order necessary? Otherwise I would prefer calling `super.viewWillLayoutSubviews` first

File: Sources/UIApplication+handleSDLEvents.swift [hunk 5]
```diff
@@ -93,6 +93,31 @@ extension UIApplication {
                 UIApplication.onWillEnterForeground()
             case SDL_APP_DIDENTERFOREGROUND:
                 UIApplication.onDidEnterForeground()
+            case SDL_WINDOWEVENT:
+                let windowEventId = SDL_WindowEventID(UInt32(e.window.event))
+
+                // seems we could also check for SDL_WINDOWEVENT_SIZE_CHANGED instead
```
Comment (michaelknoch) [inline Sources/UIApplication+handleSDLEvents.swift:99]: remove comment?

File: Sources/UIScreen+render.swift [hunk 6]
```diff
@@ -140,7 +140,7 @@ extension UIScreen {
     }
 }
 
-private extension CGRect {
+extension CGRect {
```
Comment (michaelknoch) [inline Sources/UIScreen+render.swift:143]: doesn't seem to be used, we can make this private again, no?

File: Sources/androidNativeInit.swift [hunk 7]
```diff
@@ -20,6 +20,7 @@ public func SDL_Android_Init(_ env: UnsafeMutablePointer<JNIEnv>, _ view: JavaOb
 @MainActor
 @_cdecl("Java_org_libsdl_app_SDLActivity_nativeInit")
 public func nativeInit(env: UnsafeMutablePointer<JNIEnv>, view: JavaObject) -> JavaInt {
+    print("SDLActivity_nativeInit")
```
Comment (michaelknoch) [inline Sources/androidNativeInit.swift:23]: remove print

File: Sources/androidNativeInit.swift [hunk 8]
```diff
@@ -40,6 +41,7 @@ public func nativeInit(env: UnsafeMutablePointer<JNIEnv>, view: JavaObject) -> J
 @MainActor
 @_cdecl("Java_org_libsdl_app_SDLActivity_nativeDestroyScreen")
 public func nativeDestroyScreen(env: UnsafeMutablePointer<JNIEnv>, view: JavaObject) {
+    print("SDLActivity_nativeDestroyScreen")
```
Comment (michaelknoch) [inline Sources/androidNativeInit.swift:44]: remove print

**Review by michaelknoch — COMMENTED** [review]
Looking good overall, this just needs minor cleanup

File: src/main/java/org/libsdl/app/SDLActivity.kt [hunk 9]
```diff
@@ -337,8 +337,6 @@ open class SDLActivity internal constructor (context: Context?) : RelativeLayout
             return
         }
 
-        nativeDestroyScreen()
```
Comment (michaelknoch) [inline src/main/java/org/libsdl/app/SDLActivity.kt:340]: Does moving the app in background and foreground still work?

File: Sources/UIScreen.swift (hunk 2 above)
Comment (rikner) [reply Sources/UIScreen.swift:32]: actually using .rounded here broke everything when rotating.

File: src/main/java/org/libsdl/app/SDLActivity.kt (hunk 9 above)
Comment (rikner) [reply src/main/java/org/libsdl/app/SDLActivity.kt:340]: yes it does 👍

**Review by rikner — APPROVED** [review]

**Review by michaelknoch — APPROVED** [review]

File: Sources/UIApplication+handleSDLEvents.swift [hunk 10]
```diff
@@ -93,6 +93,28 @@ extension UIApplication {
...
                 UIApplication.onDidEnterForeground()
+            case SDL_WINDOWEVENT:
+                let windowEventId = SDL_WindowEventID(UInt32(e.window.event))
+
+                guard windowEventId == SDL_WINDOWEVENT_RESIZED else {
+                    break
+                }
+
+                #if os(Android)
+                let newWidth = CGFloat(e.window.data1) / UIScreen.main.scale
```
Comment (ephemer) [inline Sources/UIApplication+handleSDLEvents.swift:104]: I think it'd be worth mentioning (in a code comment) that this matches how we init `UIScreen` on Android, due to quirks in SDL on that platform. Basically, we init UIScreen in @1x pixels on Android, but in (potentially) retina _points_ on Mac (etc.)

**Comment by ephemer** [comment]
I just had a look through the updated code and other than the code comment that I suggested we add this looks great to me 🚀
